ouroboros-network/src/Ouroboros/Network/Protocol/ChainSync/ClientPipelined.hs

```diff
@@ -0,0 +1,251 @@
+{-# LANGUAGE BangPatterns        #-}
+{-# LANGUAGE DataKinds           #-}
+{-# LANGUAGE GADTs               #-}
+{-# LANGUAGE LambdaCase          #-}
+{-# LANGUAGE NamedFieldPuns      #-}
+{-# LANGUAGE ScopedTypeVariables #-}
+
+module Ouroboros.Network.Protocol.ChainSync.ClientPipelined
+  ( ChainSyncClientPipelined (..)
+  , ClientPipelinedStIdle (..)
+  , ClientStNext (..)
+  , ClientPipelinedStIntersect (..)
+  , ChainSyncInstruction (..)
+
+  , chainSyncClientPeerPipelined
+  , chainSyncClientPeerSender
+  ) where
+
+import Network.TypedProtocol.Core
+import Network.TypedProtocol.Pipelined
+
+import Ouroboros.Network.Block (Point)
+import Ouroboros.Network.Protocol.ChainSync.Type
+
+-- | Pipelined chain sync client.  It can only pipeline 'MsgRequestNext'
+-- messages, while the 'MsgFindIntersect' are non pipelined.  This has a penalty
+-- cost of an RTT, but they are send relatively seldom and their resposne might
+-- impact how many messages one would like to pipeline.  It also simplifies the
+-- receiver callback.
+--
+newtype ChainSyncClientPipelined header point m a = ChainSyncClientPipelined {
+    runChainSyncClientPipelined :: m (ClientPipelinedStIdle Z header point m a)
+  }
+
+
+-- | Data received through pipelining: either roll forward or roll backward
+-- instcutrion.  If the server replied with 'MsgAwaitReply' the pipelined
+-- receiver will await for the next message which must come with an instruction
+-- how to update our chain.
+--
+-- Note: internal API, not exposed by this module.
+--
+data ChainSyncInstruction header tip
+    = RollForward  !header         !tip
+    | RollBackward !(Point header) !tip
```

Comment:
Isn't the same thing defined in the non-pipelined peer? Possible to re-use?

Reply:
no, there the client can use callbacks directly on received values; in the pipelined case we need to put them on a queue.